rootston/cursor.c

```diff
@@ -263,7 +294,11 @@ static void do_cursor_button_press(struct roots_input *input,
 		input->input_events_idx = (i + 1)
 			% (sizeof(input->input_events) / sizeof(input->input_events[0]));
 		set_view_focus(input, desktop, view);
-		if (view) {
+
+		struct wlr_layer_surface *layer_surface =
+			wlr_surface_layers_get_exclusive(desktop->surface_layers,
+			LAYER_SURFACE_INPUT_DEVICE_KEYBOARD);
+		if (view && !layer_surface) {
```

Comment:
This looks really strange to me. You should never be in a state where window manager focus and keyboard focus are different (window manager focus is essentially an indication of where keyboard focus is). If the surface layer accepts keyboard input (not handled here?) then it should be considered the focused window. If there is no keyboard focus, then there should be no window manager focus.

Reply:
The problem here is that surface layers are not views, thus can't have focus (and don't have the concept of focus). Also a surface layer can have exclusive keyboard but no exclusive pointer (or vice versa).

Reply:
These issues seem like they need to be addressed in the protocol.

Reply:
No, this is by design. A surface layer can choose to *receive* certain input events, and also choose to receive them *exclusively*. In the former something else will still have focus and both will get the event. The binary idea of "focus" is a bit too simple for this protocol's design.

Reply:
Oh, I didn't understood that both get the event.